numba/parfors/parfor.py

```diff
@@ -4064,28 +4072,111 @@ def remove_duplicate_definitions(blocks, nameset):
     return
 
 
-def has_cross_iter_dep(parfor):
+def has_cross_iter_dep(parfor, func_ir, typemap, index_positions=None, indexed_arrays=None, non_indexed_arrays=None):
     # we conservatively assume there is cross iteration dependency when
     # the parfor index is used in any expression since the expression could
     # be used for indexing arrays
     # TODO: make it more accurate using ud-chains
-    indices = {l.index_variable for l in parfor.loop_nests}
+    indices = {l.index_variable.name for l in parfor.loop_nests}
+    derived_from_indices = set()
+    if index_positions is None:
+        index_positions = {}
+    if indexed_arrays is None:
+        indexed_arrays = set()
+    if non_indexed_arrays is None:
+        non_indexed_arrays = set()
+
+    def add_check_position(new_position, array_accessed):
+        """Returns True if there is a reason to prevent fusion."""
+        if isinstance(new_position, list):
+            new_position = tuple(new_position)
+        if True not in new_position:
+            if array_accessed in indexed_arrays:
+                return True
+            else:
+                non_indexed_arrays.add(array_accessed)
+                return False
+        if array_accessed in non_indexed_arrays:
+            return True
+
+        npsize = len(new_position)
+        if npsize not in index_positions:
+            index_positions[npsize] = new_position
+            return False
+
+        return index_positions[npsize] != new_position
```

Comment:
I think this would benefit from some comments about what conditions are being tested and how to result ends up as True/False.

Reply:
Done.